Optimize kRelaxedSubstring LIKE pattern

xumingming opened the pull request:
kRelaxedSubstring LIKE pattern is patterns that contains '%' but it only occurs at the beginning and end of the pattern, and it also contains '_'. e.g. '%Hello_Velox%'.

The algorithm is as follows:

kRelaxedSubstring pattern is consists of two kinds of sub patterns: `kLiteralString`, `kSingleCharWildcard`, we will use notation `L0`, `Li`, `S0`, `Si` etc to describe them in the following, e.g. `<L0><S0>` is a pattern like: `%hello__%`, while `<S0><L0>` is a pattern like `%__hello%`

1. We start by set the target sub-pattern to search as `L0`(currentSubPattern = L0).
2. We search for `currentSubPattern` in the input and record the result in `currentMatchingOffsetInInput`.
3. If it is not found(`currentMatchingOffsetInInput < 0`), algorithm finish with return result `false`.
4. If it is found, we then match the right part sub-patterns
  a. If it fails at some sub-pattern `Li`, we update `currentSubPattern` as `Li`, update `currentMatchingOffsetInInput` as the next byte where we failed the matching, and jump to step 2.
  b. If it fails at some sub-pattern `Si`, algorithm finish with return result `false`.
5. If the right sub-patterns matches, we continue match left sub-patterns.
  a. If it matches, algorithm finish with return result `true`.
  b. If it failed at some sub-pattern `Lj`, we update `currentSubPattern` as `Lj`, update `currentMatchingOffsetInInput` as the next byte where we failed the matching, and jump to step 2.
  c. If it failed at some sub-pattern `Sj`, we keep `currentSubPattern` not changed, and increment `currentMatchingOffsetInInput` by one byte, and jump to step 2.

 We designed benchmarks which consists of 4 tests:

- easy: no noise in the input.
- normal: many noise in the input, but easy to have fast path.
- hard_ascii: many noise in the pure ASCII input.
- hard_unicode: many noise in the Unicode input.

Before this optimization:

```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
relaxed_substring##easy                                      1.06s   945.37m
relaxed_substring##normal                                 994.73ms      1.01
relaxed_substring##hard_ascii                                4.67s   213.91m
relaxed_substring##hard_unicode                              7.08s   141.33m
```

After this optimization:

```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
relaxed_substring##easy                                     5.73ms    174.64
relaxed_substring##normal                                   6.93ms    144.36
relaxed_substring##hard_ascii                                1.20s   836.17m
relaxed_substring##hard_unicode                              1.79s   559.71m
```

So in summary: for normal input this optimization gives us 180x speedup, for hard pure ASCII or Unicode input it gives us 4x speedup.

xumingming commented:
cc @mbasmanova

mbasmanova reviewed (commented):
@xumingming James, I'm starting to look into this optimization. Any chance you could summarize the algorithm you are using?

xumingming commented:
@mbasmanova The code has some conflicts I have to resolve first. The algorithm is as follows:

1. Start by searching the first literal sub-pattern in the input.
2. If we find the literal sub-pattern, we match the left & right part of the whole pattern.
3. If we failed matching any sub-pattern, we continue by searching the failed literal sub-pattern.

The key here is the 3rd step, we use failed sub-pattern to help us skip a lot of characters in the input.

mbasmanova commented:
@xumingming James, I'm not sure I understand step 3. Any chance you could illustrate with an example?

xumingming commented:
@mbasmanova Take the following test as an example:

```
  testLike(
      "a_ba_ba_ba_ba_ba_ba_ba_ba_ba_ba_ba_ba_ba_ba_ba_ba_b_cdef",
      "%a_b_c%",
      true);
```

The sub-patterns we are trying to match is:

- a
- _
- b
- _
- c

1. The first literal sub-pattern is `a`, so we start by searching for `a` in the input.
2. We find it at index: 0
3. We continue matching the right part of the whole pattern(_b_c), and we will fail at the sub-pattern `c`
4. Now we search for `c` in the input, we find it at the end of input.
5. Now we try to match the whole pattern at the index where we find c.
6. `c` is the right-est pattern in the whole pattern, so we no need to match the `the right part of the pattern`.
7. We match the left part of the pattern(a_b_), yeah, it matches.

We can see that by searching the failed sub-pattern `c` helped us skip a lot of characters. 

But there are cases which has a lot of noise which can not skip a lot of characters like:

```
  // There are a lot 'noise': a_b__c which is similar to our pattern: a_b_c, it
  // is hard for kRelaxedSubstring.
  testLike(
      "a_b__ca_b__ca_b__ca_b__ca_b__ca_b__ca_b__ca_b__ca_b_cdef",
      "%a_b_c%",
      true);
```

xumingming commented:
@mbasmanova Help review this one when you have time, would be great to merge it when we are still familiar with this part of code :)

mbasmanova commented on velox/functions/lib/Re2Functions.h:
```diff
@@ -101,14 +104,36 @@ class PatternMetadata {
 
-  PatternKind patternKind() const {
+  static PatternMetadata relaxedSubstring(
+      const std::string& fixedPattern,
```
Would you math other methods which take inputs by value and std::move them?

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -496,46 +496,72 @@ bool matchExactPattern(
   }
 
-  for (const auto& subPattern : patternMetadata.subPatterns()) {
+  auto pattenOffset = patternMetadata.subPatterns()[startSubPatternIndex].start;
```
typo: pattenOffset -> patternO...

nit: const

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -496,46 +496,72 @@ bool matchExactPattern(
+// Matches a kRelaxedFixed pattern forward against ASCII input.
+// Returns a pair:
+//   - first: a bool which tells us whether it matches.
+//   - second: an integer which has different meaning depends on 'first'.
```
I couldn't understand this description until I read the code. I wonder if my understanding is correct.

The result of function function might be full match, partial match, no match at all.

Full match (all sub-patterns starting from startSubPatternIndex inclusive and ending at endSubPatternIndex inclusive) is indicated by returning `{true, <index of the first character after the match>}`.

Partial match (one or more sub-patterns at the start of the list match) is indicated by returning `{false, <index of the first sub-pattern that didn't match>}`.

No match is indicated by returning `{false, -1}`.

Is this correct?

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -496,46 +496,72 @@ bool matchExactPattern(
 }
 
+// Matches a kRelaxedFixed pattern forward against Unicode input.
+// Returns a pair:
```
This description is pretty complex and is a copy-paste from above. It might be better to simply refer to above function.

```
The return values are the same as in matchRelaxedFixedForwardAscii.
```

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -557,55 +583,114 @@ std::pair<bool, int32_t> matchRelaxedFixedForwardUnicode(
-// - first: a bool indicates whether matches the pattern.
-// - second: an integer indicates where is cursor in the input when we finished
-// matching if 'first' is true, -1 otherwise.
+//   - first: a bool which tells us whether it matches.
```
Is this a 3-rd copy? Let's find a way to describe this once.

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -557,55 +583,114 @@ std::pair<bool, int32_t> matchRelaxedFixedForwardUnicode(
+        patternMetadata,
+        start,
+        startSubPatternIndex,
+        patternMetadata.numSubPatterns() - 1);
```
Do we always call matchRelaxedFixedForwardAscii and matchRelaxedFixedForwardUnicode with endSubPatternIndex equal to patternMetadata.numSubPatterns() - 1 ? If so, let's remove this argument to make the signature and calling code simpler.

mbasmanova replied:
Still wondering about this.

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -557,55 +583,114 @@ std::pair<bool, int32_t> matchRelaxedFixedForwardUnicode(
-// input use this function.
-bool matchRelaxedFixedBackwardUnicode(
+// Matches a kRelaxedFixed pattern backward against ASCII input.
+// Returns a pair:
```
more copy-paste?

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +775,97 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+  auto patternStrHeader = patternMetadata.fixedPattern().c_str();
+  // The index in the input we are matching against.
+  auto currentMatchingOffsetInInput = 0;
+  auto searchLiteralPattern = [&](SubPatternMetadata& subPattern) -> int32_t {
```
nit: pass subPattern as const &

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +775,97 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+  while (currentMatchingOffsetInInput < end) {
+    // Search current literal pattern to find out where to start matching
+    // the whole pattern.
+    currentMatchingOffsetInInput = searchLiteralPattern(currentSubPattern);
```
Is this the only place where which advance currentMatchingOffsetInInput? How do we make sure we always advance and never get stuck?

mbasmanova replied:
I'm wondering if we can advance currentMatchingOffsetInInput a bit more aggressively in some cases.

Let's say we look for sub-p N, find it at byte n, match to the right and find only partial match. I believe we can advance now to the start of the sub-pattern that failed to match and continue searching using that sub-pattern from there.

mbasmanova replied:
Still wondering about this.

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +775,97 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+    // start matching.
+    currentPatternIndex = matchResult.second;
+    currentSubPattern = patternMetadata.subPatterns()[currentPatternIndex];
+    // The mismatch pattern is not a literal means the input is not long
```
Is this happening in case of a multi-byte string? This cannot happen in case of all-ASCII string, right?

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +775,97 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+    }
+
+    // Match the left part of the pattern(if there is a left part).
+    if ((currentPatternIndex == patternMetadata.numSubPatterns() - 1) ||
```
I believe this code works, but I find it hard to follow. I wonder if there is a way to make it a bit more readable. Some ideas that come to mind:

1- replace `std::pair<bool, int32_t>` with a specific struct that uses intuitive names for fields and provides helper methods like isFullMatch(), isPartialMatch(), etc.

1- change the flow of this code to (1) match forward, check results and exit early; (2) match backward, check results.

... match forward

if (matchResult.cannotMatch()) {
   return false;
}

if (matchResult.partialMatch()) {
 ...
 continue;
}

....match backwards

if (matchResult.cannotMatch()) {
   return false;
}

if (matchResult.partialMatch()) {
 ...
 continue;
}

mbasmanova reviewed (commented):
@xumingming Overall look good, but the code is hard to read. I made some suggestions to improve readability. See if these make sense.

Last time we ended breaking production. Hence, I'll need to do more extensive testing before merging this change. Need to find time to do that. See if you can double check that test coverage is complete.

Thanks.

mbasmanova commented:
@xumingming James, would you update documentation to include this new pattern?

https://facebookincubator.github.io/velox/functions/presto/regexp.html

mbasmanova commented:
@xumingming James,

the explanation of the algorithm in https://github.com/facebookincubator/velox/pull/8581?fbclid=IwAR31kOxadJ2WzYCpDdCzmM2UzQVGOFUQAPKQwGK4GTBiFUFYalSoro73NNY#issuecomment-1925021912 is very nice and helpful. Would you include some version of it in the PR description as well?

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +775,97 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+    // Search the mismatch literal pattern to find the location where we can
+    // start matching.
+    currentPatternIndex = matchResult.second;
+    currentSubPattern = patternMetadata.subPatterns()[currentPatternIndex];
```
I'm not able to convince myself that this logic is correct when we got a full match to the right, but only partial match to the left. I keep thinking that we should either backtrack or advance one character and continue searching for the same pattern we searched before (without changing the pattern).

Any chance you can help me understand why the logic you wrote is correct?

xumingming commented:
@mbasmanova I made the following changes(all your above questions should be answered except the ones for docs):

1. Replaced std::pair<bool, int32_t> with `RelaxedPatternMatchResult` which contains `success`, `failed`, `failedForTooShortInput`, `match`, `tooShortInputToMatch` etc to make the code more easier to read.
2. Removed the duplicated comments which are now added in `RelaxedPatternMatchResult`.
3. Introduced a simple method: `isRightMostSubPattern` to make the code more readable. You suggested to change the control flow of `matchRelaxedSubstringPattern`, but I find that after all these helper method introduced, the code is simple enough, if you insist, I can try change the control flow(I think it might make the code more complex).
4. For the `right part partial match, left match failed` case, you are right, we need to add backtrack, otherwise there will be bugs, I added one case and detected it:

```c++
  // This one needs backtrack(in our implementation).
  testLike("a_b_aab_b", "%a__b%", true);
```

I added backtrack merchanism for this kind of case:

```
    // matchResult.cursor() is where we failed matching the sub-pattern, we
    // start from `matchResult.cursor() + 1` to search for the sub-pattern.
    currentMatchingOffsetInInput = matchResult.cursor() + 1;
```

Now `matchResult.cursor()` returns the cursor where we failed matching, so when matching left part of the pattern, this will give us a backtrack.

5. For your question: `Is this the only place where we advance currentMatchingOffsetInInput? How do we make sure we always advance and never get stuck?` Althrough we added backtrace which might cause the `currentMatchingOffsetInInput` going backwards, but remember that `currentMatchingOffsetInInput` is a cursor for the failed matching sub-pattern, if we looking from the first sub-pattern point of view, the cusor is always advancing, so we will not get stuck, because when search the failed sub-pattern, we are always search forward.

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -496,46 +496,119 @@ bool matchExactPattern(
+    return RelaxedPatternMatchResult{false, cursor, failingPatternIndex};
+  }
+
+  static RelaxedPatternMatchResult failedForTooShortInput() {
```
nit: failedForTooShortInput -> inputTooShort

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -496,46 +496,119 @@ bool matchExactPattern(
+    return RelaxedPatternMatchResult{false, 0, -1};
+  }
+
+  bool match() {
```
- make this and other similar methods 'const'
- also make member variables 'const'

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -496,46 +496,119 @@ bool matchExactPattern(
+    return match_;
+  }
+
+  // The input the not long enough to match.
```
typos

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -496,46 +496,119 @@ bool matchExactPattern(
             patternMetadata.fixedPattern().data() + subPattern.start,
             subPattern.length) != 0) {
-      return std::make_pair(false, -1);
+      return RelaxedPatternMatchResult::failed(currentMatchingOffset, i);
```
Perhaps, we could have a shortcut here that returns failed() if the remaining number of bytes is exactly lengthOfThePattern (input.size() - start == lengthOfThePattern)

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +796,114 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
       rows);
 }
 
+namespace {
```
Are we not already inside an anonymous namespace?

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +796,114 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+
+    // matchResult.cursor() is where we failed matching the sub-pattern, we
+    // start from `matchResult.cursor() + 1` to search for the sub-pattern.
+    currentMatchingOffsetInInput = matchResult.cursor() + 1;
```
Do we always advance only by 1 byte? I thought we have to do that only when we failed to match left side, no? Also, in case of unicode, should we advance by 1 character?

mbasmanova reviewed (commented):
RelaxedPatternMatchResult improves readability. Thank you for introduce this struct.

xumingming commented:
> @xumingming James, would you update documentation to include this new pattern?
> 
> https://facebookincubator.github.io/velox/functions/presto/regexp.html

This doc currently seems has not mentioned the patterns that we have fast path. Are you suggesting that we add one paragraph to describe the patterns we have fast path?

mbasmanova commented:
> This doc currently seems has not mentioned the patterns that we have fast path. 

I see the following: "Patterns ‘aaa’, ‘aaa%’, ‘%aaa’, where ‘aaa’ contains only regular characters and ‘_’ wildcards are evaluated without using regular expressions." We can update that sentence to include the pattern supported in this PR.

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -666,6 +783,112 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+    currentSubPattern = patternMetadata.subPatterns()[currentPatternIndex];
+    // The mismatch pattern is not a literal means the input is not long
+    // enough.
+    if (currentSubPattern.kind != kLiteralString) {
```
@xumingming Any particular reason this case is not covered by matchResult->isInputTooShort() check above?

mbasmanova commented:
> Added algorithm description in the PR description.

I see an example that shows how the algorithm works, but I don't see a formal proof that this algorithm works correctly in all cases. The description also doesn't clarify from which position we are starting next search and why.

mbasmanova commented:
see https://www.khanacademy.org/computing/ap-computer-science-principles/algorithms-101/evaluating-algorithms/a/verifying-an-algorithm

xumingming commented:
@mbasmanova 

The algorithm is as follows:

kRelaxedSubstring pattern is consists of two kinds of sub patterns: `kLiteralString`, `kSingleCharWildcard`, we will use notation `L0`, `Li`, `S0`, `Si` etc to describe them in the following, e.g. `<L0><S0>` is a pattern like: `%hello__%`, while `<S0><L0>` is a pattern like `%__hello%`

1. We start by set the target sub-pattern to search as `L0`(currentSubPattern = L0).
2. We search for `currentSubPattern` in the input and record the result in `currentMatchingOffsetInInput`.
3. If it is not found(`currentMatchingOffsetInInput < 0`), algorithm finish with return result `false`.
4. If it is found, we then match the right part sub-patterns
  a. If it fails at some sub-pattern `Li`, we update `currentSubPattern` as `Li`, update `currentMatchingOffsetInInput` as the next byte where we failed the matching, and jump to step 2.
  b. If it fails at some sub-pattern `Si`, algorithm finish with return result `false`.
5. If the right sub-patterns matches, we continue match left sub-patterns.
  a. If it matches, algorithm finish with return result `true`.
  b. If it failed at some sub-pattern `Lj`, we update `currentSubPattern` as `Lj`, update `currentMatchingOffsetInInput` as the next byte where we failed the matching, and jump to step 2.
  c. If it failed at some sub-pattern `Sj`, we keep `currentSubPattern` not changed, and increment `currentMatchingOffsetInInput` by one byte, and jump to step 2.


Let's verify this algorithm by some basic patterns first:

* For `<S0><L0>`, we start by searching `L0`(Step1, Step2), If the input does not contains `L0`, we return false(Step3). If the input contains `L0`, but the left part of input is not long enough to match `S0`(Step5c), we will continue searching `L0` in the input(Step2), and repeat util we find an index in the input where it has enough space to match `S0`, we end at Step5a; Or if there is no such index, we end at Step3.
* For `<L0><S0>`, we start by searching `L0`(Step1, Step2), If the input does not contains `L0`, we return false(Step3). If the input contains `L0`, but the right part of input is not long enough to match `S0`(Step4b), we return false as result; Otherwise we end at Step5a and return true.
* For `<S0><L0><S1>`, we start by searching `L0`(Step1, Step2), If the input does not contains `L0`, we return false(Step3). If the input contains `L0`, but the right part of input is not long enough to match `S1`(Step4b), we return false as result; If the left part of input is not long enough to match `S0`(Step5c), we will continue searching `L0` in the input(Step2), and repeat util we find an index in the input where it has enough space to match `S0` and `S1`, we end at Step5a; Or if there is no such index, we end at Step3.
* For `<L0><S1><L1>`, we start by searching `L0`(Step1, Step2), If the input does not contains `L0`, we return false(Step3). If the input contains `L0`, but can not match `L1` at the position, we will change to search for `L1`(Step4a, Step2), we might switch between `L0` and `L1` for some times, but the remaining process is similar to `<S0><L0><S1>`.

I think these four cases are the base cases, other cases is just a compound/variation of these?

xumingming commented:
For the code, I have done the following:

* Fixed some bugs -- found by formal reasoning of the algorithm
* Added tests for all the base cases -- added corresponding comments to highlight them.
* Updated the docs to include the new patterns supported.
